src/sleep/sleep.c

```diff
@@ -263,41 +263,83 @@ static int execute(
+                double time_elapsed;
 
-        log_debug("Set timerfd wake alarm for %s",
-                  FORMAT_TIMESPAN(sleep_config->hibernate_delay_sec, USEC_PER_SEC));
+                tfd = timerfd_create(CLOCK_BOOTTIME_ALARM, TFD_NONBLOCK|TFD_CLOEXEC);
+                if (tfd < 0)
+                        return log_error_errno(errno, "Error creating timerfd: %m");
 
-        timespec_store(&ts.it_value, sleep_config->hibernate_delay_sec);
+                /* Calculate current battery capacity and current time before suspension*/
+                r = read_battery_capacity_percentage();
+                if (r < 0)
+                        return log_error_errno(r, "Error fetching battery capacity percentage: %m");
 
-        r = timerfd_settime(tfd, 0, &ts, NULL);
-        if (r < 0)
-                return log_error_errno(errno, "Error setting hibernate timer: %m");
+                last_capacity = r;
+                before_timestamp = now(CLOCK_BOOTTIME);
 
-        r = execute(sleep_config, SLEEP_SUSPEND, NULL);
-        if (r < 0)
-                return r;
 
-        r = fd_wait_for_event(tfd, POLLIN, 0);
-        if (r < 0)
-                return log_error_errno(r, "Error polling timerfd: %m");
-        if (!FLAGS_SET(r, POLLIN)) /* We woke up before the alarm time, we are done. */
-                return 0;
+                log_debug("Set timerfd wake alarm to estimate battery discharge rate for %s", FORMAT_TIMESPAN(suspend_interval, USEC_PER_SEC));
+
+                timespec_store(&ts.it_value, suspend_interval);
+
+                r = timerfd_settime(tfd, 0, &ts, NULL);
+                if (r < 0)
+                        return log_error_errno(errno, "Error setting battery estimate timer: %m");
+
+                r = execute(sleep_config, SLEEP_SUSPEND, NULL);
+                if (r < 0)
+                        return r;
+
+                r = fd_wait_for_event(tfd, POLLIN, 0);
+                if (r < 0)
+                        return log_error_errno(r, "Error polling timerfd: %m");
+
+                if (!FLAGS_SET(r, POLLIN))  { /* We woke up before the alarm time, calculate battery discharged and estimate rate. */
+                        after_timestamp = now(CLOCK_BOOTTIME);
+
+                        time_elapsed = (after_timestamp - before_timestamp) / USEC_PER_SEC;
+
+                        r = read_battery_capacity_percentage();
+                        if (r < 0)
+                                return log_error_errno(r, "Error fetching battery capacity percentage: %m");
+
+                        current_capacity = r;
+                        log_debug("%d%% battery discharged in %f seconds time",
+                                        last_capacity - current_capacity,
+                                        time_elapsed);
+
+                        estimated_discharge_rate = ((last_capacity - current_capacity) * 60 * 60) / time_elapsed;
```

Comment:
so, while the system was sleeping the user might have plugged in a power sources, and unplugged it again, so we never know, maybe after the sleep the battery might be *more* charged than before. We must handle that in some way.

Reply:
Sure. Will fix that.

Reply:
Is there a way to detect that AC might have been plugged and un-plugged ? Or i can simply apply filter to calculate estimate_discharge_rate only when current_capacity is less than last_capacity. Does that work ?